---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# MR LOVAR (Latent Outcome Variable Approach) R correlation between snps

<!-- badges: start -->
<!-- badges: end -->

**`MRLOVAR`** is an R package designed to enhance Mendelian Randomization (MR) analysis by incorporating a latent phenotype of the outcome, which enables a more precise disentanglement of horizontal and vertical pleiotropy effects. This feature allows for a more explicit assessment of the exclusion restriction assumption and iteratively refines causal estimates using the expectation-maximization algorithm. Compared to existing MR methods, this approach offers a unique and potentially more accurate framework. particularly when genetic variants are correlated, and this correlation is estimated from reference populations such as the 1000 Genomes Project (1kG).

In addition to its core functionality, the package includes permutation testing, providing a non-parametric method for hypothesis testing that controls the type 1 error rate under the null hypothesis without relying on specific distributional assumptions. 

**`MRLOVA`** also features two additional statistical tests:

-  **Directional Pleiotropy Test:**  Assesses whether genetic variants influence the outcome through pathways other than the exposure of interest.
-  **InSIDE Assumption Test:** Evaluates whether the assumption that Instrument Strength Independent of Direct Effect (InSIDE) holds.

There are three functions:

-  **`mr_lovar()`**  main function that performs MR analysis and provides causal effect estimate.
-  **`InSIDE()`**  performs Instrument Strength Independent of Direct Effect (InSIDE) assumption test.
-  **`directional()`**  performs directional pleiotropy test.

To view the help pages for functions in this package, prepend the function name with a question mark.

```{r}
#library(MRLOVA)
#?mr_lova
#?InSIDE
#?directional
```
## Installation

You can install the development version of **MRLOVA** from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("lamessad/MRLOVA")
```
## Steps for Mendelian Randomization Analysis with MRLOVA
**1. Load Summary Statistics:**

-  Begin by loading the summary statistics for the instrumental variables.

**2. Harmonize Data:**

-  Check the alleles and effects between the exposure and outcome studies.
-  Flip the sign of `betaY` if the effect allele in the exposure study differs from the effect allele in the outcome study.

**3. Standardize data:**

-  Standardize the summary statistics before performing the MR analysis. The summary statistics are standardized as z-statistics and then rescaled by the sample size for both continuous and binary traits.

**4. Standardize data:**

-  Estimate the correlation between the IVs using the reference population. Match and reorder the correlation matrix according to GWAS summary data.

**5. Perform MR Analysis:**

-  Use the standardized and harmonized summary statistics to perform the Mendelian Randomization analysis.

**Note**: If you are using the **“TwoSampleMR”** package, you can leverage the harmonized summary data to create the input parameters for `MRLOVA` as follows:

-  **Continuous outcome** - the summary statistics are standardized as z-statistics and then rescaled by the sample size.

-  **Binary outcome** - If the effect size of the outcome GWAS summary data was reported in Odds Ratio (OR), convert the OR to log(OR). Then, the summary statistics are standardized as z-statistics and then rescaled by the sample size.

``` r
#library(TwoSampleMR)
#library(MRLOVAR)
#dat=harmonize_data(exposure_data, outcome_data)
#detaX=dat$z.exposure/sqrt(dat$samplesize.exposure)
#betaY=dat$z.outcome/sqrt(dat$samplesize.outcome) 
#betaXse=1/sqrt(dat$samplesize.exposure)
#betaYse=1/sqrt(dat$samplesize.outcome)
#ny=dat$samplesize.outcome
#ld_cor=cor_matrix(path_plink2, bfile, dat)
#est = mr_lovaR(betaY, betaX, betaYse, betaXse, ld_cor, ny, permutn = 1000,log_file = "log.txt")  
```

## Example

Here is an example demonstrating how to apply MRLOVAR methods to infer the causal effect from exposure to outcome, assuming the first two steps mentioned above have been carried out.


```{r example}
#MR analysis
library(MRLOVAR)
data("dat")
data("ld_cor") 
head(dat)
dat$nx=dat$ny # sample size of exposure
betaX = dat$betaX/dat$betaXse/sqrt(dat$nx)
betaY = dat$betaY/dat$betaYse/sqrt(dat$ny)
betaXse = 1/sqrt(dat$nx)
betaYse = 1/sqrt(dat$ny)
ny = dat$ny
est = mr_lovaR(betaY, betaX, betaYse, betaXse, ny, permutn = 1000) 
est
```

#### Warnings
The package includes built-in warnings to help ensure the robustness of your analysis. Please be aware of the following conditions and corresponding warnings:

**Convergence Warnings**
-  If the number of iterations is less than 3 or greater than 10, the package will issue a warning:

```r
#warning("Please check convergence.")
#log_message("Warning: Please check convergence.", log_file)
```
This indicates potential issues with the convergence process. Details of the convergence are saved in the `log.txt` file in the directory where you are running the analysis. Reviewing this log file can provide insights and help troubleshoot potential problems.

**Permutation Warnings**
-  Performing 1000 permutations is typically sufficient for most analyses. However, for a more accurate p-value, increasing the number of permutations is recommended. Ideally, the number of permutations should exceed `1 / causal effect p-value`, although this can be impractical. If the number of permutations is less than `1/est$CausEstP` (where `est$CausEstP` is the estimated p-value without permutation), the package will issue a warning:

```r
#warning("#To get a more precise p-value, it is recommended to increase the number of permutations, given the p-value of causal effects  =  ", est$CausEstP)
#log_message(paste("# To get a more precise p-value, increasing the number of permutations is recommended, given the p-value of causal effects, p-value of causal effect  =  ", est$CausEstP), log_file)
```
Here are two outputs of the permutation analysis: `est$sig_v`(the 5th percentile of the permuted p-values) and `corrected_p`(the corrected p-value based on the permutation distribution). The permuted p-values form an empirical null distribution under the assumption of no effect. The 5th percentile of this distribution provides a threshold or critical value for significance. If `est$CausEstP` is less than `est$sig_v`,  it indicates that the observed p-value is significantly smaller than what would be expected under the null hypothesis. The `corrected_p` is the proportion of permuted p-values that are less than or equal to the observed p-value (`est$CausEstP`), representing the corrected p-value based on permutation.

### InSIDE

**1. Using all instrumental variables:**

```{r}
# InSIDE assumption test
inside_all = InSIDE(betaY, betaX, as.numeric(est$CausEst))  
inside_all
```

**2. Using only valid instrumental variables:** The `MRLOVA` package identifies valid instrumental variables, and their indices are listed in the analysis output as `est$Valid`.

```{r}
# InSIDE assumption test
inside = InSIDE(betaY[est$Valid], betaX[est$Valid], as.numeric(est$CausEst))  
inside
```

### directional

**1. Using all instrumental variables:**

```{r}
#directional pleiotropy test
dir_pleiotropy=directional(betaY, betaX, as.numeric(est$CausEst))
dir_pleiotropy
```

**2. Using only valid instrumental variables:**

```{r}
#directional pleiotropy test
dir_pleiotropy=directional(betaY[est$Valid], betaX[est$Valid], as.numeric(est$CausEst))
dir_pleiotropy
```

**Note**: This novel statistical test has more statistical power than the MR-Egger intercept for detecting directional pleiotropy, while also controlling the Type I error rate(refer the manuscript).

## Contact

Please contact Lamessa Amente (lamessa.amente@mymail.unisa.edu.au) or Hong Lee (hong.lee@unisa.edu.au) if you have any queries.
